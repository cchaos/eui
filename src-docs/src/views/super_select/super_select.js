import React, {
  Component,
  Fragment,
} from 'react';

import {
  EuiSuperSelect,
  EuiSpacer,
  EuiText,
  EuiIcon,
} from '../../../../src/components';

export default class extends Component {
  constructor(props) {
    super(props);

    this.options = [
      {
        value: 'option_one',
        inputDisplay: 'Option one',
        dropdownDisplay: (
          <Fragment>
            <strong>Option one</strong>
            <EuiSpacer size="xs" />
            <EuiText size="s" color="subdued">
              <p className="euiTextColor--subdued">Has a short description giving more detail to the option.</p>
            </EuiText>
          </Fragment>
        ),
      },
      {
        value: 'option_two',
        inputDisplay: (
          <span>
            <EuiIcon type="alert" /> Option two
          </span>
        ),
        dropdownDisplay: (
          <Fragment>
            <strong>Option two</strong>
            <EuiSpacer size="xs" />
            <EuiText size="s" color="subdued">
              <p className="euiTextColor--subdued">Has a short description giving more detail to the option.</p>
            </EuiText>
          </Fragment>
        ),
      },
      {
        value: 'option_three',
        inputDisplay: 'Option three has a super long text to see if it will truncate or what',
        dropdownDisplay: (
          <Fragment>
            <strong>Option three</strong>
            <EuiSpacer size="xs" />
            <EuiText size="s" color="subdued">
              <p className="euiTextColor--subdued">Has a short description giving more detail to the option.</p>
            </EuiText>
          </Fragment>
        ),
      },
    ];

    this.state = {
      value: this.options[1].value,
    };
  }

  onChange = (value) => {
    this.setState({
      value: value,
    });
  };

  render() {
    return (
      <Fragment>
        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          hasDividers
          aria-label="Use aria labels when no actual label is in use"
        />

        <EuiSpacer size="m" />

        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          disabled
          aria-label="Use aria labels when no actual label is in use"
        />

        <EuiSpacer size="m" />

        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          isLoading
          aria-label="Use aria labels when no actual label is in use"
        />

        <EuiSpacer size="m" />

        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          isLoading
          disabled
          aria-label="Use aria labels when no actual label is in use"
        />

        <EuiSpacer size="m" />

        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          compressed
        />
      </Fragment>
    );
  }
}
