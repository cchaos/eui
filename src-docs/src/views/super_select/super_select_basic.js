import React, {
  Component,
  Fragment,
} from 'react';

import {
  EuiSuperSelect,
  EuiHealth,
} from '../../../../src/components';

export default class extends Component {
  constructor(props) {
    super(props);

    this.options = [
      {
        value: 'warning',
        inputDisplay: (
          <EuiHealth color="subdued">
            Warning
          </EuiHealth>
        ),
      },
      {
        value: 'minor',
        inputDisplay: (
          <EuiHealth color="warning">
            Minor
          </EuiHealth>
        ),
      },
      {
        value: 'critical',
        inputDisplay: (
          <EuiHealth color="danger">
            Critical
          </EuiHealth>
        ),
      },
    ];

    this.state = {
      value: this.options[1].value,
    };
  }

  onChange = (value) => {
    this.setState({
      value: value,
    });
  };

  render() {
    return (
      <Fragment>
        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          aria-label="Health levels"
        />
      </Fragment>
    );
  }
}
