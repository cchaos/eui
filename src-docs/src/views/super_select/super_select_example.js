import React from 'react';

import { renderToHtml } from '../../services';

import {
  GuideSectionTypes,
} from '../../components';

import {
  EuiCode,
  EuiSuperSelect,
} from '../../../../src/components';

import SuperSelect from './super_select';
const superSelectSource = require('!!raw-loader!./super_select');
const superSelectHtml = renderToHtml(SuperSelect);

export const SuperSelectExample = {
  title: 'SuperSelect',
  sections: [{
    title: 'SuperSelect',
    source: [{
      type: GuideSectionTypes.JS,
      code: superSelectSource,
    }, {
      type: GuideSectionTypes.HTML,
      code: superSelectHtml,
    }],
    text: (
      <p>
        Description needed: how to use the <EuiCode>EuiSuperSelect</EuiCode> component.
      </p>
    ),
    props: { EuiSuperSelect },
    demo: <SuperSelect />,
  }],
};
