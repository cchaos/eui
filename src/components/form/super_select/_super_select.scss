/*
 * 1. Make popover the same width as the form control
 * 2. Style popover similar to combobox
 */

.euiSuperSelect.euiPopover { /* 3 */
  @include euiFormControlSize; /* 1 */
  height: auto;

  .euiPopover__anchor {
    display: block;
  }
}

.euiSuperSelect__popoverPanel[class*="bottom"] {
  border-top-color: transparentize($euiBorderColor, .2);
  border-top-right-radius: 0; /* 2 */
  border-top-left-radius: 0; /* 2 */
}

.euiSuperSelect__popoverPanel[class*="top"] {
  border-bottom-color: transparentize($euiBorderColor, .2);
  border-bottom-right-radius: 0; /* 2 */
  border-bottom-left-radius: 0; /* 2 */
  @include euiBottomShadowFlat; /* 2 */
}

.euiSuperSelect.euiPopover--anchorDownCenter.euiPopover-isOpen .euiSuperSelect__popoverPanel { /* 3 */
  transform: translateX(-50%) translateY(0) translateZ(0); /* 2 */
}

.euiSuperSelect__item--hasDividers:not(:last-of-type) {
  border-bottom: $euiBorderThin;
}
