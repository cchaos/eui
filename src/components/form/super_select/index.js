export {
  EuiSuperSelect,
} from './super_select';
