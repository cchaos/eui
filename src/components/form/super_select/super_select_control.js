import React, { Fragment } from 'react';
import { renderToString } from 'react-dom/server';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {
  EuiFormControlLayout,
} from '../form_control_layout';

export const EuiSuperSelectControl = ({
  className,
  options,
  id,
  name,
  inputRef,
  fullWidth,
  isLoading,
  hasNoInitialSelection,
  defaultValue,
  compressed,
  value,
  ...rest
}) => {
  const classes = classNames(
    'euiSuperSelectControl',
    {
      'euiSuperSelectControl--fullWidth': fullWidth,
      'euiSuperSelectControl--compressed': compressed,
      'euiSuperSelectControl-isLoading': isLoading,
    },
    className
  );

  let emptyOptionNode;
  if (hasNoInitialSelection) {
    emptyOptionNode = (
      <option value="" disabled hidden style={{ display: 'none' }}>&nbsp;</option>
    );
  }

  // React HTML input can not have both value and defaultValue properties.
  // https://reactjs.org/docs/uncontrolled-components.html#default-values
  let selectDefaultValue;
  if (!value) {
    selectDefaultValue = defaultValue || '';
  }

  let selectedValue;
  if (value) {
    const selectedOption = options.find(option => option.value === value);
    selectedValue = selectedOption.inputDisplay;
  }

  const icon = {
    type: 'arrowDown',
    side: 'right',
  };

  // In the case the inputDisplay is a react element, make it a string, then strip the HTML
  // This is used to read only the text for a screenreader.
  const selectedValueAsString = renderToString(selectedValue).replace(/<\/?[^>]+(>|$)/g, '');

  return (
    <Fragment>
      <select
        id={id}
        name={name}
        className="euiSuperSelectControl__hiddenField"
        ref={inputRef}
        defaultValue={selectDefaultValue}
        value={value}
        aria-hidden="true"
        {...rest}
      >
        {emptyOptionNode}
        {options.map((option, index) => {
          const {
            text, // eslint-disable-line no-unused-vars
            value,
          } = option;
          return <option value={value} key={index} />;
        })}
      </select>

      <EuiFormControlLayout
        icon={icon}
        fullWidth={fullWidth}
        isLoading={isLoading}
        compressed={compressed}
      >

        <button
          role="option"
          type="button"
          className={classes}
          aria-haspopup="true"
          aria-label={`Select an option: ${selectedValueAsString} selected`}
          {...rest}
        >
          {selectedValue}
        </button>

      </EuiFormControlLayout>
    </Fragment>
  );
};

EuiSuperSelectControl.propTypes = {
  name: PropTypes.string,
  id: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.node.isRequired,
    inputDisplay: PropTypes.node.isRequired
  })).isRequired,
  isInvalid: PropTypes.bool,
  fullWidth: PropTypes.bool,
  isLoading: PropTypes.bool,

  /**
   * Simulates no selection by creating an empty, selected, hidden first option
   */
  hasNoInitialSelection: PropTypes.bool,
  inputRef: PropTypes.func,
  /**
   * when `true` creates a shorter height input
   */
  compressed: PropTypes.bool,
};

EuiSuperSelectControl.defaultProps = {
  options: [],
  fullWidth: false,
  isLoading: false,
  hasNoInitialSelection: false,
  compressed: false,
};
