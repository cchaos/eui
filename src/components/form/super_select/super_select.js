import React, {
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EuiSuperSelectControl } from './super_select_control';
import { EuiPopover } from '../../popover';
import { EuiContextMenuItem } from '../../context_menu';
import { keyCodes } from '../../../services';

const SHIFT_BACK = 'back';
const SHIFT_FORWARD = 'forward';

export class EuiSuperSelect extends Component {
  constructor(props) {
    super(props);

    this.itemNodes = [];
    this.state = {
      isPopoverOpen: props.isOpen || false,
    };
  }

  setItemNode = (node, index) => {
    this.itemNodes[index] = node;
  };

  openPopover = () => {
    this.setState({
      isPopoverOpen: true,
    });

    const focusSelected = () => {
      const indexOfSelected = this.props.options.reduce(
        (indexOfSelected, option, index) => {
          if (indexOfSelected != null) return indexOfSelected;
          if (option == null) return null;
          return option.value === this.props.valueOfSelected ? index : null;
        },
        null
      );

      // valueOfSelected is optional, and options may not exist yet
      if (indexOfSelected != null) {
        // wait for the CSS classes to be applied, removing visibility: hidden
        requestAnimationFrame(() => this.focusItemAt(indexOfSelected));
      } else {
        requestAnimationFrame(focusSelected);
      }
    };

    requestAnimationFrame(focusSelected);
  };

  closePopover = () => {
    this.setState({
      isPopoverOpen: false,
    });
  };

  itemClicked = (value) => {
    this.setState({
      isPopoverOpen: false,
    });
    this.props.onChange(value);
  };

  onSelectKeyDown = e => {
    if (e.keyCode === keyCodes.UP || e.keyCode === keyCodes.DOWN) {
      e.preventDefault();
      e.stopPropagation();
      this.openPopover();
    }
  }

  onItemKeyDown = e => {
    if (e.keyCode === keyCodes.ESCAPE) {
      // close the popover and prevent ancestors from handling
      e.preventDefault();
      e.stopPropagation();
      this.closePopover();
    } else if (e.keyCode === keyCodes.TAB) {
      // no-op
      e.preventDefault();
      e.stopPropagation();
    } else if (e.keyCode === keyCodes.UP) {
      e.preventDefault();
      e.stopPropagation();
      this.shiftFocus(SHIFT_BACK);
    } else if (e.keyCode === keyCodes.DOWN) {
      e.preventDefault();
      e.stopPropagation();
      this.shiftFocus(SHIFT_FORWARD);
    }
  }

  focusItemAt(index) {
    const targetElement = this.itemNodes[index];
    if (targetElement != null) {
      targetElement.focus();
    }
  }

  shiftFocus(direction) {
    const currentIndex = this.itemNodes.indexOf(document.activeElement);
    let targetElementIndex;

    if (currentIndex === -1) {
      // somehow the select options has lost focus
      targetElementIndex = 0;
    } else {
      if (direction === SHIFT_BACK) {
        targetElementIndex = currentIndex === 0 ? this.itemNodes.length - 1 : currentIndex - 1;
      } else {
        targetElementIndex = currentIndex === this.itemNodes.length - 1 ? 0 : currentIndex + 1;
      }
    }

    this.focusItemAt(targetElementIndex);
  }

  render() {
    const {
      className,
      options,
      valueOfSelected,
      onChange,
      isOpen,
      hasDividers,
      itemClassName,
      itemLayoutAlign,
      ...rest
    } = this.props;

    const buttonClasses = classNames(
      {
        'euiSuperSelect--isOpen__button': this.state.isPopoverOpen,
      },
      className,
    );

    const itemClasses = classNames(
      'euiSuperSelect__item',
      {
        'euiSuperSelect__item--hasDividers': hasDividers,
      },
      itemClassName,
    );

    const button = (
      <EuiSuperSelectControl
        options={options}
        value={valueOfSelected}
        onChange={onChange}
        onClick={this.state.isPopoverOpen ? this.closePopover : this.openPopover}
        onKeyDown={this.onSelectKeyDown}
        className={buttonClasses}
        {...rest}
      />
    );

    const items = options.map((option, index) => {
      return (
        <EuiContextMenuItem
          key={index}
          className={itemClasses}
          icon={valueOfSelected === option.value ? 'check' : 'empty'}
          onClick={() => this.itemClicked(option.value)}
          onKeyDown={this.onItemKeyDown}
          layoutAlign={itemLayoutAlign}
          buttonRef={node => this.setItemNode(node, index)}
        >
          {option.dropdownDisplay || option.inputDisplay}
        </EuiContextMenuItem>
      );
    });

    return (
      <EuiPopover
        className="euiSuperSelect"
        panelClassName="euiSuperSelect__popoverPanel"
        button={button}
        isOpen={isOpen || this.state.isPopoverOpen}
        closePopover={this.closePopover}
        panelPaddingSize="none"
        anchorPosition="downCenter"
        ownFocus={false}
      >
        {items}
      </EuiPopover>
    );
  }
}

EuiSuperSelect.propTypes = {
  /**
   * Classes (and `...rest`) will be applied to the control
   */
  className: PropTypes.string,
  /**
   * Classes for the context menu item
   */
  itemClassName: PropTypes.string,
  /**
   * You must pass an `onChange` function to handle the update of the value
   */
  onChange: PropTypes.func,
  /**
   * Pass an array of options that must at least include:
   * `value`: storing unique value of item,
   * `inputDisplay`: what shows inside the form input when selected
   * `dropdownDisplay` (optional): what shows for the item in the dropdown
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      inputDisplay: PropTypes.node,
      dropdownDisplay: PropTypes.node,
    }),
  ).isRequired,
  valueOfSelected: PropTypes.string,
  /**
   * Change to `true` if you want horizontal lines between options.
   * This is best used when options are multi-line.
   */
  hasDividers: PropTypes.bool,
  /**
   * Change `EuiContextMenuItem` layout position of icon
   */
  itemLayoutAlign: PropTypes.string,
};

EuiSuperSelect.defaultProps = {
  options: [],
};
