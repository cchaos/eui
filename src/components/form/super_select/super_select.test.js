import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test';

import { EuiSuperSelect } from './super_select';

describe('EuiSuperSelect', () => {
  test('is rendered', () => {
    const component = render(
      <EuiSuperSelect {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
