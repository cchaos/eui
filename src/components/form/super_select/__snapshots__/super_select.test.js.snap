// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSuperSelect is rendered 1`] = `
Array [
  <select
    aria-hidden="true"
    aria-label="aria-label"
    class="euiSuperSelect__hiddenField"
    data-test-subj="test subject string"
  />,
  <div
    class="euiFormControlLayout"
  >
    <button
      aria-label="aria-label"
      class="euiSuperSelect testClass1 testClass2"
      data-test-subj="test subject string"
    />
    <div
      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
    >
      <span
        class="euiFormControlLayoutCustomIcon"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs>
            <path
              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
              id="arrow_down-a"
            />
          </defs>
          <use
            fill-rule="nonzero"
            href="#arrow_down-a"
          />
        </svg>
      </span>
    </div>
  </div>,
]
`;
